### PR TITLE
Deprecate/Add projection methods

### PR DESCRIPTION
**Pull request by jburel**
# What this PR does

This PR allows the projection along T.
This PR introduces new methods so the projection axis can be specified
This PR deprecates methods not specifying the projection axis

# Testing this PR

Integration tests have been added to test:
 *  the projection itself
 * the rendering of projection preview
 * comparison tests between projection preview and render of projected image

Check that the ``RenderingEngineTest`` and ``ProjectionServiceTest`` are still green.

Not covered in this PR are tests to check if the projection algorithm (outside the scope of the PR)

for example:
 * retrieve bytearray for each z-section.
 * do a maximum intensity projection looping through the arrays
 * compare with the result of a maximum intensity projection using the projection method from the ``ProjectionBean`` 

# Related reading

https://trello.com/c/JkRcxqVU/96-z-projection-thumbnails

### CODE REVIEW

File: components/blitz/resources/omero/Constants.ice [hunk 1]
```diff
@@ -296,6 +296,26 @@ module omero {
         MEANINTENSITY,
         SUMINTENSITY
       };
+
+      /**
+       * The projection axis
```
Comment (joshmoore) [inline components/blitz/resources/omero/Constants.ice]: What are the intended uses for the constant strings as opposed to the enum?

Comment (jburel) [reply components/blitz/resources/omero/Constants.ice]: This is an old commit (august) and might no longer be needed.
Thanks for pointing that out

**Comment by jburel** [comment]
No test failures:
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-integration-java/lastCompletedBuild/testngreports/integration/ProjectionServiceTest/
and 
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-integration-java/lastCompletedBuild/testngreports/integration/RenderingEngineTest/

**Comment by joshmoore** [comment]
A design question: This moves the API from v1 to v2 methods by adding a new enum parameter. Is there a potential for needing future additions? Would something more extensible like a command object be preferrable?

**Comment by jburel** [comment]
I don't think we will need other options until we work on "n-dimension"
The concept of modulo is already tricky enough!

File: components/blitz/resources/omero/api/IProjection.ice [hunk 2]
```diff
@@ -135,12 +192,85 @@ module omero {
...
+                 *             to the Pixels qualified by
+                 *             <code>pixelsId</code> will be used with a
+                 *             <i>Projection</i> suffix. For example,
+                 *             <i>GFP-H2B Image of HeLa Cells</i> will have an
+                 *             Image name of
+                 *             <i>GFP-H2B Image of HeLa Cells Projection</i>
+                 *             used for the projection.
+                 * @return The Id of the newly created Image which has been
+                 *         projected.
+                 * @throws ValidationException Where:
```
Comment (mtbc) [inline components/blitz/resources/omero/api/IProjection.ice]: maybe lowercase "where"

File: components/server/src/ome/services/RenderingBean.java [hunk 3]
```diff
@@ -552,22 +553,13 @@ public RGBBuffer render(PlaneDef pd) {
...
             throw new ResourceError(e.getMessage());
         } finally {
             rwl.writeLock().unlock();
-            try {
-                if (byteStream != null) {
-                    byteStream.close();
-                }
-            } catch (IOException e) {
-                log.error("Could not close byte stream.", e);
-                throw new ResourceError(e.getMessage());
```
Comment (mtbc) [inline components/server/src/ome/services/RenderingBean.java:569]: it's okay to drop the wrapping in a `ResourceError`?

File: components/server/src/ome/services/projection/ProjectionBean.java [hunk 4]
```diff
@@ -115,12 +131,27 @@ public void setPixelsService(PixelsService pixelsService)
...
+                    ctx.from = pixelBuffer.getStack(channelIndex, plane);
+                default:
+                    break;
+                case IProjection.T_AXIS:
+                    //retrieve the hypercube
+                    List<Integer> offset = Arrays.asList(new Integer[]{
+                            0, 0, plane, channelIndex, 0});
+                    List<Integer> size = Arrays.asList(new Integer[]{
+                            ctx.pixels.getSizeX(),
+                            ctx.pixels.getSizeY(), 1, 1,ctx.pixels.getSizeT()});
```
Comment (mtbc) [inline components/server/src/ome/services/projection/ProjectionBean.java]: space after comma

File: components/tools/OmeroJava/test/integration/ProjectionServiceTest.java [hunk 5]
```diff
@@ -572,4 +573,98 @@ public void testProjectImageByGroupOwnerRWRW() throws Exception {
...
+        int channel = 0;
+        int step = 1;
+        IProjectionPrx svc = factory.getProjectionService();
+        try {
+            svc.projectPlanes(pixels.getId().getValue(),
+                    null, ProjectionType.MAXIMUMINTENSITY,
+                    ProjectionAxis.MODULOT, z,
+                    channel, step, 0, pixels.getSizeT().getValue()-1);
+            Assert.fail("An error should have been thrown");
+        } catch (Exception e) {
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ProjectionServiceTest.java]: maybe can be more specific about which? `ServerError` or somesuch?

**Comment by mtbc** [comment]
Are we planning tests of the along-t projection algorithm results for 5.3.0? Or will we soon be adding to graphical client so we can kind of eyeball it?

File: components/server/src/ome/services/RenderingBean.java (hunk 3 above)
Comment (jburel) [reply components/server/src/ome/services/RenderingBean.java:569]: I did not want to change the behaviour of the existing method i.e. ``projectPixels``

**Comment by jburel** [comment]
@mtbc: I don't think we will have time to implement the projection along T in the UI for 5.3.0
That was not the plan

**Comment by mtbc** [comment]
So is there a way in which we will think by 5.3.0 that the projection along T offered by this PR does appear to do the projection correctly?

File: components/server/src/ome/services/RenderingBean.java (hunk 3 above)
Comment (mtbc) [reply components/server/src/ome/services/RenderingBean.java:569]: Ah, sorry, I had thought this was changing the behavior, my mistake.

**Comment by dominikl** [comment]
Looks good. Integration test ran and passed, too. 👍

**Comment by jburel** [comment]
@mtbc @dominikl 
See https://github.com/openmicroscopy/openmicroscopy/pull/5057
Tomorrow build should allow to visually inspect the projection along T

**Comment by mtbc** [comment]
All seems good to me too, thank you. :+1:

**Pull request closed without merging**